### PR TITLE
Add generic retry support for kinesis consumers

### PR DESCRIPTION
**Pull request by jlogeart**

### CODE REVIEW

File: app/io/flow/event/v2/KinesisConsumer.scala [hunk 1]
```diff
@@ -116,14 +120,35 @@ case class KinesisRecordProcessor[T](
...
+        if (retries > MaxRetries) {
+          logger_
+            .withKeyValue("retries", retries)
+            .error(s"[FlowKinesisError] Error while processing records after $MaxRetries", e)
+          throw e
+        } else {
+          logger_
+            .withKeyValue("retries", retries)
+            .warn(s"[FlowKinesisWarn] Error while processing records (retry $retries/$MaxRetries). Retrying...", e)
+          executeRetry(records, retries + 1)
```
Comment (nataliab) [inline app/io/flow/event/v2/KinesisConsumer.scala]: Let's add `Thread.sleep(BackoffTimeInMillis)` between the retries.

File: app/io/flow/event/v2/KinesisConsumer.scala [hunk 2]
```diff
@@ -116,14 +120,35 @@ case class KinesisRecordProcessor[T](
...
 
+  @tailrec
+  private def executeRetry(records: Seq[Record], retries: Int): Unit = {
+    Try(f(records)) match {
+      case Success(_) =>
+      case Failure(NonFatal(e)) =>
+        if (retries > MaxRetries) {
+          logger_
+            .withKeyValue("retries", retries)
+            .error(s"[FlowKinesisError] Error while processing records after $MaxRetries", e)
```
Comment (nataliab) [inline app/io/flow/event/v2/KinesisConsumer.scala]: Should this be reported as `error` or `warn`? See Mike's comment on https://github.com/flowcommerce/lib-event-relation-mapper/pull/121#pullrequestreview-216224939

Comment (nataliab) [reply app/io/flow/event/v2/KinesisConsumer.scala]: Also, should we be throwing exception or just logging with `warn`?

File: app/io/flow/event/v2/KinesisConsumer.scala [hunk 3]
```diff
@@ -116,14 +120,35 @@ case class KinesisRecordProcessor[T](
       )
     }
 
-    f(flowRecords)
+    executeRetry(flowRecords, 0)
```
Comment (gheine) [inline app/io/flow/event/v2/KinesisConsumer.scala]: Should retry be a value >0

File: app/io/flow/event/v2/KinesisConsumer.scala (hunk 2 above)
Comment (gheine) [reply app/io/flow/event/v2/KinesisConsumer.scala]: Would it make sense to delay the retry exponentially? In any case I'd recommend using `Retry` in gfc-util here.

Comment (jlogeart) [reply app/io/flow/event/v2/KinesisConsumer.scala]: - Mike's comment makes sense but the change is consistent with `handleCheckpoint`
 - Fine with swallowing and logging, explicitly stating that the records are skipped.
 - I suggest keeping a flat waiting time for now.

File: app/io/flow/event/v2/KinesisConsumer.scala (hunk 3 above)
Comment (jlogeart) [reply app/io/flow/event/v2/KinesisConsumer.scala]: This is more like a counter of retries, indicate the current retry and not the remaining.

**Pull request merged**